### ISSUE TITLE
Ternary operator should not be more complex (npath) than a if/else condition.

### ISSUE THREAD

**Issue by h0gar** [pdepend/pdepend#158]
```
    function test($a) {
          return $a > 10 ? 5:20;
    }
```

has only 2 possible paths, but pdepend gives it a npath complexity of 5.

See the discussion at: http://stackoverflow.com/questions/24197973/why-is-the-ternary-operator-more-complex-than-if-else

**Comment by hakre** [comment]
http://www.literateprogramming.com/mccabe.pdf

**Comment by h0gar** [comment]
Thanks for the link but... quick summary please? And I only had a quick glance but it doesn't seem to mention ternary operators..

**Comment by yoosefi** [comment]
Nobody can seem to justify this behavior. :laughing:

**Comment by hakre** [comment]
It's easily justified: As existing algorithms calculate this way and it's a creating a number to process further on with it, you're never intended to change it. For that matter, no justification is necessary, even if you're completely not able to understand the rationale behind.

**Comment by yoosefi** [comment]
"no justification is necessary"
"can't change without breaking something"

These have no place in a discussion about theory. Sorry, try again.

**Comment by hakre** [comment]
Mouth full of theory but can't read.

**Comment by yoosefi** [comment]
Pedantic attitude but can't cite.

**Comment by h0gar** [comment]
> It's easily justified: As existing algorithms calculate this way and it's a creating a number to process further on with it, you're never intended to change it (otherwise you break the metric which is counter-productive as you have a metric here). For that matter, no justification is necessary, even if you're completely not able to understand the rationale behind.

You mean it's just some sort of backward compatibility for former metrics. Well, old metrics can probably be updated, especially when they are based on a bug. I'm not sure that's a valid reason for not fixing it.

**Comment by hakre** [comment]
So far you've only see the crack in the matrix, but you speak as if you've taken the pill already.

**Comment by hakre** [comment]
In a sidenote: Sure things could be updated, but apart from bug or not and even if a bug, it's most likely not worth. The metric does work equally well. The difference can be neglected. Unless your code is full of ternary operators but if so, I'm pretty sure project management already handled it well with the numbers it generated.

**Comment by h0gar** [comment]
> So far you've only see the crack in the matrix, but you speak as if you've taken the pill already.

uh?! ok...

> even if a bug

it's a bug. I have explained it 10 times already.

> The metric does work equally well. The difference can be neglected.

It's not. Cyclomatic complexity is exponential, so having 5 instead of 2 makes a big difference. A function with 5 conditions has 2^5 npath, but 5^5 with ternary operators. 32 or 3125 is not a "neglectable" difference.

**Comment by hakre** [comment]
The npath complexity is academic. I really see no reason to change that. And yes I've understood you feel this is a bug and I respect your feelings. But unless you point out how the implementation in pdepend differs from the specification I can't see where the flaw is nor how to go towards some change in the code. Just calling it a bug doesn't change the specs it has to follow.

Some of the metrics also count the number of operators. I don't know if this is the case for that metric, but there can be some formal reasons why it's calculated that way. I'm no academic and not very much into these metrics all that deep but even if you've got one anomaly here or there, it's metrics, you shouldn't have that many ternary operators that it actually makes such a difference (even if you looking on the isolated issue see such a sheer difference in the numbers), it's not wise to put that much focus on little details that don't deserve so much. Even wrong.

**Comment by Rarst** [comment]
> you shouldn't have that many ternary operators that it actually makes such a difference

Speaking for practical experience of running phpmd integration in my IDE — it makes huge difference very fast. As above — it's exponential.

I don't think people are using these tools for pure academic joy of it. I am using it to see a complexity of code I am working with and keep it low. If the tool gives me skewed results for the sake of academic purity of implementation or whatever — it is useless for my practical purposes.

**Comment by hakre** [comment]
Well I actually would consider the "keeping it low" academic joy. Those metrics should be part of larger calculations you use, e.g. cost estimates for the ongoing development. If you only look on the numbers w/o context most likely on isolated methods or a single module you will only see exponentiality on the number as it's calculated exponential. I can perfectly understand that these breakouts in metrics really bothers programmers, but I personally think as a coder, I should not take these little details that seriously concerning writing software.

This is only a metric. As far as I can trust the author he has implemented it correctly - which is very important otherwise you couldn't use it as metric (under name/title "Npath Complexity").

On the other hand there seem to be some users which are upset by high numbers and now struggle with an anomaly within the named metric. But as much as any model you use in your designs can't show the whole picture, the metric can not either.

And from all who criticize the implementation, it was never argued with the specification.

You could however suggest *Rarsts PHP Burst Npath Complexity* specification documentation and then to implement such a metric within your toolchain. I don't see any problem with that as long as it means some use to you. Funnily this (or similar) has never been suggested here. It's all about moaning there would be a flaw in the implementation without pointing to the part of the specification of the metric where it has been violated. But insisting it would have a bug.

If the original specification has a mistake, you would need to fix that upstream first as well. So you can't have just an issue here w/o the academic work to do to battle the original specs.

I'm not saying it's free of mistakes, I'm just missing the point here for a valid suggestion on how to productively deal with the dissent. If I would be the author of pdepend I would do the hell to rewrite the implementation of the Npath Complexity metric calculation just because some developers (and it really could be me, I don't remember but there might have been a time I might have been bothered by this, too) on the internet are really concerned and are upset about the bug that exists in general Npath Complexity calculation when applied onto PHP code.

**Comment by Rarst** [comment]
I am not qualified to comment on spec, and I hadn't _seen_ actual spec posted _in the open_ yet (I think the only link was to a article behind paywall).

However it was stated in related thread:

> the fact that ternary gives an artificially higher complexity of 5 vs the correct result of 2 is a bug. It's 5 because of an internal detail of Java.

> https://github.com/phpmd/phpmd/issues/124#issuecomment-81957256

I get that there are different ways to look at this. In my opinion it's not worth it to elevate implementation quirks in specific language to an immutable language–agnostic things–are–done–this–way–because spec status.

**Comment by hakre** [comment]
I can see that, but anthony has not provided reference for the claim that this is a detail of java (yet).

He perhaps is capable of doing so, so it's great to have him in, but still the implementation doesn't look with quirks to me:

- ternary operator: https://github.com/pdepend/pdepend/blob/master/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php#L189
-  if construct: https://github.com/pdepend/pdepend/blob/master/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php#L254

Also from a less academic perspective: Why not listen to the guts? Isn't it that ternary operators tend to be complex when read? More than the if-else clause? Doesn't look to me the complexity metric would be awfully wrong with that.

**Comment by yoosefi** [comment]
> So far you've only see the crack in the matrix, but you speak as if you've taken the pill already.

This is laughably condescending. You have yet to "enlighten us" or even demonstrated your own understanding of it, so I don't think you have any justification in speaking this way.

This metric is not "purely academic", people use this to identify problematic areas in their production code. The artificially high N-path is a false flag and indicates that the tool is *broken*.

>  Isn't it that ternary operators tend to be complex when read? More than the if-else clause?

That is not how path complexities are determined.

**Comment by hakre** [comment]
> That is not how path complexities are determined.

I'd say this is what it has been made for. Complexity of code has been identified as quite a useful metric (for let's say compared with LOC). I'd even say it's the one aiming to measure the complexity you normally feel in your guts and it has the benefit that your guts sometimes lie to you (especially when you're confronted to defend your work) the metrics won't do. Not that I praise for numbers here, it's just one measurement.

Why I brought that point up is that from non-academic review (somebody brought up the point he or she is too lazy to read and want's to be spoon fed with the lecture) it actually can make sense to have a different (higher) number for different (more) complexity. If that premise would already be broken then the implementation most likely would be more subject to suspicion.

So it would be useful if those in opposition to the implementation could share their personal experience if they see (in sense of a personal sensing or feeling, not in what you believe how it should be calculated) the ternary operator more, less or equally complex than an if/else clause. What do *you* think?

As for the complexity calculation, the graph of the ternary operator has not been visualized in the paper I linked (I suppose there was none in [PDP-10](https://en.wikipedia.org/wiki/PDP-10) FORTRAN?). But the IF THEN ELSE control structure has been visualized by McCabe:

![image](https://cloud.githubusercontent.com/assets/378849/6900399/77813004-d706-11e4-88b7-7139a5054435.png)

For cyclomatic complexity:

> the maximum number of linearly independent circuits in G*

The mathematical preliminaries needed for the definition/theorem of the *cyclomatic number* V(G) can be found in *C. Berge, Graphs and Hypergraphs. Amsterdam, The  Netherlands: North-Holland, 1973*.

> *Definition 1:* The cyclomatic number V(G) of a graph G with n vertices, e edges, and p connected components is

> v(G) = e - n + p.

> *Theorem 1:* In a strongly connected graph G, the cyclomatic number is equal to the maximum number of linearly independent circuits.

On the matter of that p = 1 I have not read that far yet, the remark is:

> The role of the variable p will be explained in Section IV .For these examples assume p = 1.

This can be read up all in the beginning of McCabe. What I found interesting is the point around *linearly independent circuits*. So it's perhaps not as easy as just seeing the graph but to also take note on the number of exit nodes and circles within.

I do know that in *NPATH: a measure of execution path complexity and its applications* it is related to McCabe (Reference 11) but as I don't have the text I can't say to which level. I do assume it's on the matter of this said *cyclomatic number*  and the calculation of *cyclomatic complexity* and that McCabe created the base of the calculation Nejmeh makes use of. But for some reason they also added some extras and as I don't have the document at hand I can't say to which part of the calculation of cyclomatic complexity or the measurement of the cyclomatic number this breaks down.

So I'd say if there is to battle how NPATH complexity calculation has been outlined for ternary operators (which I think this report should be about), one could start the argument finding out how McCabe would have measured the complexity for it. That's perhaps done best black-box without knowing about Nejmeh.

Another approach could be to find a flaw in the implementation directly. I've already linked the parts of the code in question but I've not seen anyone putting focus on it yet. Because what I theoretically thought of is that perhaps the implementation has not been done right.

I take the position to defend the implementation (which is really easier as it's the status quo). So some of you guys has to make the counterpart. Sorry if that is more work for you (but I actually provided some docs at least, so you can't say I'm making it easy for myself).

Even I would prefer to see the path and measurement and calculation rebutted per what McCabe did outline, I'd also like to see more details at which level you think the implementation is wrong. Ircmaxell has brought up the point that this is fault due to a Java detail but he hadn't had the opportunity to explain what he means. I've not seen anyone else making that point.

And generally I'd say: If you think it's a bug, you have to provide more than just that per your own logik both have the same path. As the implementation differs I would at least be that respectful to take the difference as an opportunity for you even you battle the implementation you can explain why it's a bug by the book. That is how the path would look like (as in theory), how it would be measured and how it would be calculated.

I'm pretty sure there is a reason to make a difference between how if/else is calculated and how the ternary operator are dealt with. I wrote it before, there might be a distinction between a control structure and operators.

**Comment by yoosefi** [comment]
@hakre if you scroll up I gave some code which I think would justify the ternary paths resulting in 5 after it is expanded. All I am looking for is whether or not my example is indeed what is going on. This is however not *supposed* to be how the ternary is interpreted. The ternary is supposed to be a shorthand if/else and expand to that.

**Comment by hakre** [comment]
All I see is that you changed some code. I do not see how that is following McCabe. Which was what I think should be done. Not just repeating that it must have been if/else while as we know it's not treated that way (but yet the reason is unclear).

**Comment by yoosefi** [comment]
As I've cited above, other analyzers correctly show the McCabe of ternary is 2.

**Comment by hakre** [comment]
I beg your pardon, as far I've seen you've linked *one* analyzer (http://ars.altervista.org/lint_php/lint_php.php), not more than one. [ref](https://github.com/pdepend/pdepend/issues/158#issuecomment-57895698) 

Also that analyzer does not really deal with ternarys. From what I could find from a very quick view on it is that it converts ternaries to if/else structures. What we've found out so far is that we can't use if/else and ternary 1:1 for complexity. But that script does that.

It's a bit of a pity that version control is lost and we can't better review the dates in .2009 when ternary was patched in. This is lost from public history. Not saying this code is pure shit, I haven't looked much in it, just saying it doesn't cut.

As you've been written about many analysers can you point to another one?

**Comment by yoosefi** [comment]
the js one

**Comment by hakre** [comment]
@yoosefi. I've not seen it earlier, thanks for pointing out. Can you reference the implementation for the ternary of the JS one?

**Comment by yoosefi** [comment]
Same code, dude. `x ? true : false`

**Comment by hakre** [comment]
@yoosefi: You mistunderstood it. I mean the code where the analysis of the *ternary* is done. Implementation please.

**Comment by yoosefi** [comment]
But as you've said, black box is preferable.

Also
>I do assume it's on the matter of this said cyclomatic number and the calculation of cyclomatic complexity and that McCabe created the base of the calculation Nejmeh makes use of.

Not gonna entertain an argument based on assumptions. Give me the proof that says McCabe treated ternary as different.

Hint: He didn't.

Ternary is shorthand if-then-else. As the graph you cited points out, this should be 2 paths.

**Comment by yoosefi** [comment]
I'm wondering what you think the new `??` operator in PHP 7 should be. If you think `5` I'm interested in what frankenlogic you use to justify that.

**Comment by hakre** [comment]
Oh, how well you deliver, looks like you're cherry-picking low hanging fruits to make your claims but when asked for actual things like code or docs, you're shedding away. How unexpected!

And you must be missing something from this confrontation: I don't have to give proof. I have brought up the argument, that when you want to have changes in the implementation provided with Pdepend you have to override the document Pdepend implemented (which is from Nejmeh btw. and not McCabe but you must hate precision it seems).

Again it was me who provided the assumption that such document is based on McCabe (as a service to opponent the implementation which I don't have to do) but I also outlined that - despite having McCabe not read in full yet (and I was always open about that matter) - it is that McCabe didn't talk about ternary operators at all because the computer language he is writing about does not have ternary-operators. So I can't give proof that McCabe treats ternary different from the point alone that McCabe didn't even talk about ternary - and that could have been clear for you already before asking, so no news at all and absolutely the expected outcome that there will be no proof. If you really expected otherwise, please blame your ignorance on that matter.

Not even possible but this proof would be negligible because it's not to disproof McCabe here but Nejmeh. I wrote that if you could bring up that with McCabe it might make it easier to battle Nejmeh for you then (that is *for you*, still you have to bring up the arguments and as you asked for proof, how coincident, that now means, you have to bring that exact proof. The proof for Nimjeh has been [already provided by Manuel](https://github.com/pdepend/pdepend/issues/158#issuecomment-55022816)).

You really create the impression that you're lazy to really put some substance to your claim. Even more serious, you seem to have missed so far even what your claim actually is: From the little quote that is available above from Nejmeh's paper we're not talking about the number of paths in Nejmeh or McCabe but the number *added* (this is *perhaps* what p is in McCabe but I don't know as I didn't read Nejmeh and McCabe only in part yet). But you still bark the wrong tree here that everybody (than you) is confabulating a different number of paths which is just not the case here.

So just to make this clear again, like with the proof you have asked for: 

I can not see any dissent across all documents (to the extend of which have been referenced excerpts in full or in part from here openly) *and* within the discussions here so far **about the number of paths**. And this information is available since pretty much the beginning. Sorry to point out the obvious here.

If you don't care to actually trying to explain why what you observed is different to your subjective conception, well then I guess it makes no sense at all to relate to any kind of scientific papers but science normally is about to finding exactly that out. And yes, it's quite some work, but you perhaps think you get logic for free but if not, hey, it's way cheaper to blame somebody else who takes a different opinion. Which I for that matter didn't do.

So you can still bark on me like a little but loud dog, or just look into the matter and start writing about it instead of just moaning because a number you got was unexpected. I can honour this as an initial event for you taking part in this discussion, but after some time you have to come further than just these initial feelings if you want to see progress here. Because otherwise nobody would really care.

But I perhaps wrote too much: Proof Nimjeh wrong. That is your claim. Now proof it.

**Comment by yoosefi** [comment]
>McCabe didn't talk about ternary operators at all because the computer language he is writing about does not have ternary-operators.

Correct.

>The proof for Nimjeh has been already provided by Manuel.

That document is behind a paywall.

> words words words

lol

**Comment by yoosefi** [comment]
You have yet to address the `??` operator.

**Comment by hakre** [comment]
@yoosefi: The whole document is behind a paywall, that's true, but the relevant part has been quoted. By your reaction you're again showing that you're absolutely lazy and you have no serious intention to progress the issue you brought up. And you still don't get it: I don't have to address the `??` operator at all. I don't even have to bring up the ternary. I just service reflection of your comments. You're the suppliant.

**Comment by yoosefi** [comment]
lol ok dude.

well, by your citation, if-then-else has 2 paths. thanks for agreeing with me and @h0gar

**Comment by hakre** [comment]
Absolutely, that's what I tried to tell you. So you need to understand the rest of the equation of that metric. Which requires (I assume) understanding McCabe but effectively and finally Nimjeh. Luckily Java has a ternary operator like PHP so Nimjeh could specify the calculation in his work. If I have read that right, the 2 results in the actual "+ 2" at the end of the formula he wrote.

**Comment by yoosefi** [comment]
if-then-else = 2

if-then-else = 5

thanks for playing

**Comment by hakre** [comment]
Well @yoosefi I can't do the reading for you, that's at least something you must do your own. But you can continue to neglect that and bark the wrong tree. I can only point you to the places where the measurements are done, which calculations are used and what the reasoning behind it is. All that information is available in the thread here.

**Comment by yoosefi** [comment]
you haven't demonstrated anything. you've used a lot of words to effectively say nothing.

**Comment by hakre** [comment]
@yoosefi it's not my job nor my intention to entertain you with fancy demonstrations and it never was my intention. All I wrote is that you need to do the reading your own. Now that must have been worded vaguely, but rest assured  I meant less my limited words, but just the source-code and the papers written by others. Those texts you refuse to even discuss. From what you share about your cognition, it ends with the finding of numbers you don't understand. Something you don't understand must be illogical. I can personally accept that for you, it's just that in my opinion this is a bit limited and not very thoughtful for a discussion on a topic with disagreements.

**Comment by yoosefi** [comment]
@h0gar and i are trying to have a fruitful discussion and you keep trying to shut it down by either saying "that's just the way things are" or "well we just don't know so get over it".

please, be fruitful and answer the question about the `??` operator, which is slated to be very popular in php 7.

**Comment by yoosefi** [comment]
For Java, please see http://pmd.sourceforge.net/pmd-4.3.0/xref/net/sourceforge/pmd/rules/design/NpathComplexity.html#228

You will see on line 239 that npath simply has 2 added to it. This verifies that the nodes are counted as separate functions and not part of an inline construct, as I stated here https://github.com/pdepend/pdepend/issues/158#issuecomment-56925098

Again, we are back to the question: Why is this the case, why is this not commented on in the source, and why has this been ported over?

**Comment by hakre** [comment]
@yoosefi: I don't think that *"counted as separete functions"* describes it well. It's said that the *"2 that is included in the NP(?) expression reflects the execution path complexity resulting from this statement" (stated here: https://github.com/pdepend/pdepend/issues/158#issuecomment-55022816). 

> Why is this the case, 

From the information presented, this implementation follows Nimjeh.

> why is this not commented on in the source, and why has this been ported over?

It's commented in the source (base of calculation) and it has been additionally commented here. It should be at least be clear by now.

The sources you present form java are even additionally. Thanks for that link.

So if I read you right, the question then is: Why did Nimjeh differ between if/else and ternary operator? More correct?

**Comment by hakre** [comment]
And about `??` my answer with my current limited understanding is NP(?). And you're perhaps right with what you called a function, this is probably what McCabe writes about in "FLOW" and subroutines vs. control structures. While the first are used to break the program apart into blocks, the other part is in calculating the complexity of those blocks.

Now if if/then/else creates blocks for further flow analysis but the ternary constitutes a single block that is being calculated, this could explain the differentiation.

Compare with *III. WORKING EXPERIENCE WITH THE
COMPLEXITY MEASURE* in McCabe, the first two paragraphs.

**Comment by yoosefi** [comment]
NP(?) being 2, right?

Not sure what you mean as "separate blocks". Can you produce some code that demonstrates what you think ternary is being expanded to?

**Comment by hakre** [comment]
McCabe distinguishes in the code analysis between *blocks* that are extracted in the *flow* [perhaps I make this too short here, this is more my own current mental model]. Those blocks are then calculated. That is for the if/else structure the path of the structure is not taken into account (as that are separate blocks) whereas for the ternary there is no structure, just one block so you have the path complexity count here. This is how I read McCabe now and this makes sense to me, it's perhaps just the viewpoint from what you're looking but you can't compare both complexity calculations on paths only because they are structurally differently applied. if/else is more an inner complexity calculation and ternary more an outer one. Perhaps that's an image to explain. That's how I read this now.
And for NP(?), here is the calculation again:

> NP(?) = NP(<expr1>) + NP(<expr2>) + NP(<expr3>) + 2

**Comment by yoosefi** [comment]
> if/else is more an inner complexity calculation and ternary more an outer one

That's what I suggested in my first reply to this thread. However, McCabe didn't make this distinction, and Nemjeh didn't show his work as far as I can tell (no graph cited here, and I can't find one on Google). So, this is "Nemjeh's magical complexity" and not McCabe complexity. Can you produce a graph of ternary where E - N = 5?

**Comment by yoosefi** [comment]
I'd additionally like to cite what Eclipse IDE does, which treats it as an if/else (page 15): http://eclipsefcg.sourceforge.net/Documentation.pdf

**Comment by yoosefi** [comment]
Can you please produce a graph of this "special if/then" with its 5 acyclical paths?

**Comment by hakre** [comment]
No, I'm not good in producing graphs. But apart from that, I don't see any dissent about the graph in the whole discussion.

**Comment by yoosefi** [comment]
To prove that ternary has more paths than if/else.

**Comment by yoosefi** [comment]
> NP(?) = NP(<expr1>) + NP(<expr2>) + NP(<expr3>) + 2

The problem with this is that `NP(<expr1>)` is double-counting the parent path, which must evaluate `<expr1>` before the branch. `+2` is then double counting `<expr2>` and `<expr3>`.

So, the correct calculation would be, the same as if/else:

> NP(?) = NP(<then>) + NP(<else>)

**Comment by hakre** [comment]
I think the number of paths (2) are only added as a complexity calculation result. So there are not 5 paths but 2.

I don't know if Nejmeh has a proof in that article either. If you're concerned about the paywall, you can try to reach the author personally and ask for a copy of his original article, not the derivative work the publisher has reproduced. In the end that technically speaks about two different documents, but I don't think there are many differences.

**Comment by yoosefi** [comment]
CC and NPath are separate things. They shouldn't be added together.

**Comment by hakre** [comment]
If that is true, then most likely this is not CC and NPath added together here. Compare with McCabe. Or do you want to say now, that McCabe is wrong, too?

**Comment by yoosefi** [comment]
McCabe for if/else as reported by PDepend is 2. Which is correct. The NPath is incorrect. That is what this entire ticket is about. Have you not been paying attention?

**Comment by yoosefi** [comment]
Kind of tired of talking to you. It's like talking to a wall. Still waiting on that graph, dude.

**Comment by yoosefi** [comment]
Honestly can you provide no better argument than "some dude wrote it in an article 30 years ago" ?

**Comment by jpirie** [comment]
Thought I'd just leave a comment here as I've been thinking about this today, and don't see a great answer to this question above. Consider this code block `if (exp) { exp2 } then { exp3 }`. This has an NPath complexity of 3, that's one for evaluating `exp`, one for the evaluation of `exp2` and one for the evaluation of `exp3`

However, `exp ? exp2 : exp3` gives an NPath complexity is 5. We have the same calculations as before, but we have to add an additional 2 to the NPath because we can do value assignments with the ternary form, and so we gain this additional 2 from doing the return values of `exp2` and `exp3`. To clarify, it's because we can do `String x = true ? "test" : "testing"`. We do not have to consider return in a standard `if` form, because `String x = if (true) { ... } {...}` is syntactically invalid.

I hope this helps the next person who comes along.

**Comment by yoosefi** [comment]
@jpirie I'm sorry, but that is incorrect, assignments can also be done inside the branches. The ternary is a branch shorthand. Could you please provide a graph of these 5 paths?

**Comment by hakre** [comment]
Just to add a third persons view to it: I could find literature of this exact code-complexity topic within a title of Tom de Marco in the 80's (I only have the german translation so I quote german, I hope this is not a too heavy problem for anyone involved).  He writes about McCabe (1976) and also Chen (1978). For the graph he writes:

> Die Methode wird dadurch am besten klar, wenn ein Stück Code mit seinen Verbindungen durch eine Art Flussdiagramm dargestelt wird (Abbildung 12.4).
> [Abbildung 12.4 als Bild; Bildunderschrift: Abbildung 12.4: Flussgraph, der das Modul UNIQUE der Abbildung 12.2 darstellt.]
> Die Knotenpunkte dieser Abbildung stehen für Codeabschnitte, die aufeinander folgend ausgeführt werden müssen (es gibt keine Verzweigungspunkte innerhalb dieser Knotenpunkte). Die Linien stehen für Kontrollwege, die diese Abschnitte verbinden (Abbildung 12.5).

My quotation misses the graph picture, but it's easily described: Just fairly large, solid, black dots connected with thin lines which have a thin arrow-head at one end. As the code-example of the UNIQUE example it depicts is not part of the quotation as well, there is not much more to say to get a good description on how the graph looks like: solid dots connected with thin arrows. 

Following this description (the graph model Tom de Marco writes about in his title), I was able to sketch the following graphs for the ternary operator:

![path8084-1-a](https://cloud.githubusercontent.com/assets/378849/8442551/f856cfd8-1f7f-11e5-934f-b4888a047741.png)

I call it *the butterfly*. Regarding Chen this illustration is misleading, a lower number can be obtained when painted like this:

![path8084-1-b](https://cloud.githubusercontent.com/assets/378849/8442556/fd274b96-1f7f-11e5-9d9c-e43b67ff3635.png)

So not only the implementation of Pdepend as well as in the Java world, but already Chen has a higher complexity number (compared to the graph of an if/else code example). It is Tom de Marco who writes that Chen did the empiric verification of McCabe. Taking this as granted, the complexity as measured by the book is higher for a code block of a ternary operator than code blocks of an if/else structure. The graph(s) show it.

**Comment by yoosefi** [comment]
Your graph describes a loop. And there are only 2 paths.

Please provide a graph with 5 paths.

**Comment by yoosefi** [comment]
The translated german is:

```
The method is thus best understood if a piece of code is dargestelt with its connections by a kind flowchart ( Figure 12.4 ) .
[ Figure 12.4 as an image ; Image Under Font : Figure 12.4 : Flow graph illustrating the UNIQUE module of Figure 12.2 ] .
The nodes of this figure represent sections of code that must be executed sequentially ( there are no branch points within these hubs ) . The lines represent control paths that connect these sections ( Figure 12.5 ) .
```

Why even include this at all? It has no explanation whatsoever.

**Comment by yoosefi** [comment]
Also, related to my question earlier regarding the Null coalescing operator `??`,

https://en.wikipedia.org/wiki/Null_coalescing_operator

> In contrast to the ternary conditional if operator used as x ? x : y, the null coalescing operator is a binary operator and thus evaluates its operands at most once, which is significant if the evaluation of x has side-effects.

So, `??` does not branch. It has the same complexity as any other binary operator such as `&&` or `||`

**Comment by yoosefi** [comment]
So, after a year of discussion, and no explanation in sight, I've reached this conclusion:
=====

This metric is broken. Don't use ternary if you rely on this metric.

You can somewhat cheat, by not using branches at all. Here is a comparison of NPaths for various functions that do the same thing.

```
<?php

// npath 1
// ccn 1
function squashed ( $x ) {
    $x && $x = 1 or $x = 2;
    return $x;
}

// npath 5
// ccn 2
function ternary ( $x ) {
    return $x ? 1 : 2;
}

// npath 2
/ ccn 2
function ifelse ( $x ) {
    if ($x) return 1;
    else return 2;
}
```

I can't even begin to imagine a directed acyclical graph with 5 paths for single-level branch, and despite openly requesting one multiple times, nothing has been produced, and the justification for this behavior is to just shrug it off.

**Comment by hakre** [comment]
... not to forget to shrug it off by suggesting to write code differently because you've got the opinion a metric might be wrong. This is really getting hilarious.

So a caution-note for your health: Please don't fall-apart by investing so much investigations on the issue. I mean, if I let the last year pass again, you've provided *soooo much* references here, I wasn't even able to read them. The number of days on which you provided references within the last year divided by the number of pages in references you provided just leaves me with a division by zero.

**Comment by yoosefi** [comment]
Please provide a graph with 5 paths.

**Comment by yoosefi** [comment]
Didn't think so.

**Comment by AlexisFinn** [comment]
Wow, just happened upon this "issue" myself, and I was wondering if it was some kind of flaw or not.
So it seems it's a "flaw" in the npath complexity calculation.
ok.
I might want to point out that these numbers are only indicators poiting you to a potentially overly complex code.
But in the end it's up to the developper to decide wether the code truly needs to be changed or not.

Here's an example function I have in a controller that validates the input from a pretty simple form:
```php
private function validateForm($input)
    {
        $errors = array();

        $lastName = isset($input['lastName']) ? trim($input['lastName']) : null;
        if (!Zend_Validate::is($lastName, 'StringLength', array(1, 100))) {
            $errors['lastName'] = $this->__('Le nom n\'est pas valide');
        }

        
        $firstName = isset($input['firstName']) ? trim($input['firstName']) : null;
        if (!Zend_Validate::is($firstName, 'StringLength', array(1, 100))) {
            $errors['firstName'] = $this->__('Le prénom n\'est pas valide');
        }

        $email = isset($input['email']) ? trim($input['email']) : null;
        if (!Zend_Validate::is($email, 'EmailAddress')) {
            $errors['email'] = $this->__('L\'adresse mail n\'est pas valide');
        }

        if (isset($input['sms'])) {
            
            $phoneNumber = isset($input['phoneNumber']) ? $input['phoneNumber'] : null;
            //country will also be verified in validatePhoneNumber by the lib
            $country = isset($input['country']) ? $input['country'] : null;
            try {
                $validator = new Validator();
                $validator->isPhoneNumberValid($phoneNumber, $country);
            } catch (Exception $e) {
                $errors['phoneNumber'] = $this->__($e->getMessage());
            }
        }

        if (isset($input['hasLegacy'])) {
            $legacy = isset($input['legacy']) ? $input['legacy'] : null;
            if (!Zend_Validate::is($legacy, 'NotEmpty')) {
                
                $errors['legacy'] = $this->__('Pour poursuivre, vous devez lire et accepter les conditions générales d\'utilisation');
            }
        }

        return $errors;
```

This has an Npath complexity of a whooping 561000 !!!
Could it be re-factored to lower that complexity: yes, certainly.
But does it need to be refactored ? Not really

The function is not too long, it has a clearly defined purpose, it's not difficult to understand, debug or modify if necessary. 
So basiacally,if you have some time to spare, it would probably be a good idea to refactor the validation of each input into a separate function, but that astronomically large NPath complexity can pretty safely be ignored .

**Comment by yoosefi** [comment]
@AlexisFinn this effectively makes the npath hueristic, an otherwise very useful indicator of problematic areas, useless. The real bug I think is the dev simply handwaving to some whitepaper behind a paywall and snubbing anyone that disagrees. Someone should fork this project and fix it.

**Comment by yoosefi** [comment]
Furthermore I'm interested in how he's going to implement the null coalescence operator. It contributes no branching, just like other logical operators, but I will not be surprised if he makes it have an npath of 5 just like ternary out of laziness.

**Comment by yoosefi** [comment]
I like @hakre 's (non-sarcastic, unironic) explanation that we simply don't understand because we're n00bs trapped in the matrix and haven't taken the red pill yet. Scientific rigor and open source development attitude at its finest.

**Comment by Radio** [comment]
@manuelpichler, The formula is correct but shouldn't the NP((exprX)) give 0 for cases when the expression is just a constant or a variable? So it should still be: 0 + 0 + 0 + 2 = 2 as for if/else ?

**Comment by yoosefi** [comment]
https://github.com/pdepend/pdepend/blob/master/src/main/php/PDepend/Metrics/Analyzer/HalsteadAnalyzer.php#L252
```
/*
* The ternary operator ‘?’ followed by ‘:’ is considered a
* single operator as it is equivalent to “if-else” construct.
*/
```
Really now?

@Radio his "formula" is to always add 3. Out of thin air. 

https://github.com/pdepend/pdepend/blob/master/src/main/php/PDepend/Metrics/Analyzer/CyclomaticComplexityAnalyzer.php#L415

```
        ++$data[self::M_CYCLOMATIC_COMPLEXITY_1];
        ++$data[self::M_CYCLOMATIC_COMPLEXITY_2];
```

**Comment by bryanagee** [comment]
Now reddit has taken up your fight =).

https://www.reddit.com/r/programming/comments/42cf4n/on_researching_some_wacky_cyclomatic_complexity/

**Comment by dolkensp** [comment]
KumbajaMyLord on Reddit has actually identified the bug that is causing this - and yes, it is a bug, even according to the spec:

https://www.reddit.com/r/programming/comments/42cf4n/on_researching_some_wacky_cyclomatic_complexity/cz9cvgl

@yoosefi @hakre

**Comment by pygy** [comment]
For future reference, in case  that thread was to disappear:

>This is so sad.

>If you look past the petty argument and actually look at the code of PDepend it is so obvious that this is a bug. The code maintainer 'manuelpichler' is correct and the cited paper is correct. The actual bug is what the (as of now) last poster 'Radio' posted. The NPATH for the subexpressions in the ternary operator are calculated incorrectly, but no one in that shitty flame war focuses on that.

>For what it's worth, this is the particular piece of code: ([src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php around line 213](https://github.com/pdepend/pdepend/blob/b93462cf520be6660d9fa5768b6daf565b0f47fd/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php#l214-l216))
```PHP
    foreach ($node->getChildren() as $child) {
        if (($cn = $this->sumComplexity($child)) === '0') {
            $cn = '1';
        }
        $npath = MathUtil::add($npath, $cn);
    }
```

>The implementation goes out of its way to make the NPATH complexity of each subexpression at least 1, which is simply incorrect. The NPATH of an expression should be the number of && and || operators and since a constant or variable has 0 of these operators their NPATH should be 0.

**Comment by kbilsted** [comment]
Bug or no bug. A large part of the problem with this discussion is the lack of understanding that if-else is different to the ternary operator.  The ternary operator is an *expression* whereas the if-else is a *statement*.

You can say (in Java )

```
int daysOfWork = isboss () ? 2 : 20;
```

You can not say

```
int daysOfWork = if ( isboss ()){ 2 } else  { 20 };
```

So the two constructs are different and thus *may* entail different complexity calculations.

**Comment by h0gar** [comment]
I don't understand how this could matter for the complexity. Whether it's an expression or a statement, the number of paths is still the same.

**Comment by lesliev** [comment]
According to the abstract: http://dl.acm.org/citation.cfm?id=42379
`NPATH ... counts the acyclic execution paths through a function`

Doesn't seem like any mystery to me. Both ternary and IF statements have just two possible execution paths. @h0gar @yoosefi why not just fork this and fix it? Time to fix: 5 minutes. Time to argue: years.

**Comment by dolkensp** [comment]
@kbilsted - tell that to Icon...

x[if y > 5 then 5 else y] := "Y"

Anyways, expressions are also statements, by definition - ternary operators are just more restricted if statements. They don't somehow become more complex, by virtue of being more restrictive. If anything, the potential for cyclomatic complexity goes DOWN.

**Comment by yoosefi** [comment]
Please supply a graph with five acyclical paths.

**Comment by yoosefi** [comment]
I stand corrected.
![](https://i.imgur.com/mXMSsYB.png)

**Comment by kbilsted** [comment]
@h0gar  *expressions* can be used for expressing things not directly possible with statements, eg.

```
const string boo =

**Comment by yoosefi** [comment]
@kbilsted 
> "not directly possible"

```
if (isboss()) daysofwork = 2;
else daysofwork = 20;
```
and
```
if (isboss()) print array[2];
else print array[1];
```

**Comment by yoosefi** [comment]
@lesliev according to you, in the time you took to comment that you could have forked and fixed this.

**Comment by h0gar** [comment]
@kbilsted npath measures the complexity of the algorithm. It doesn't matter if you use expression or statement. Only the interpreter cares about it. In the end, the algorithm is still exactly the same. You could use plain english words instead of statements/expressions/whatever, it would still be the same algorithm.

@lesliev yeah. Someone on reddit was already trying to make a patch. I don't need it anymore so I'm not feeling in a hurry but I could give it a try if the other guy can't make it.

**Comment by yoosefi** [comment]
![](https://i.imgur.com/bGyWuuf.jpg)

>I'm going to give you five choices, Neo.

**Comment by lesliev** [comment]
@yoosefi maybe, if I was a PHP programmer... as it is I'm just a nosey bystander. I appreciate that the bug might be difficult to fix, though I can't imagine that fixing it would take more than a year. PS: love your graph :grinning:

**Comment by kbilsted** [comment]
@yoosefi you can jump up and down  all day it still doesn't   make a statement an expression. 

 Also you conveniently rewrote my examples to have a different meaning.. hence not understanding at all my point.

**Comment by yoosefi** [comment]
@kbilsted please provide a graph with 5 acyclical paths

**Comment by dolkensp** [comment]
@kbilsted You need to get back on track.

The tool is meant to measure the complexity of an algorithm, not just add random numbers together based on the overlying construct.

Are you telling me that

    function myCheck($a, $b, $c, $d) {
        if ($a < $b)
            return true;
        else if ($a > $c)
            return false;
        else if ($c < $d)
            return false;
        return true;
    }
    
    ...
    
    if (myCheck(2,4,6,8))
    {
        return "Hi";
    }
    else
    {
        return "Bye";
    }

has less branches than this

    return 2 < 4 ? "Hi" : "Bye";

Because right now, you're claiming that the SYNTAX matters more than the actual BRANCHING which is what this tool is meant to be measuring.

Get off your high horse, admit defeat, and stop championing an unrelated syntax war to cover your original mistake.

Perhaps, in future, you could develop https://github.com/kbilsted/pnumstatements or https://github.com/kbilsted/pnumexpressions - in which case, we'll all come to your aid if you get in this argument again - but for now, you are defeated. Go home.

**Comment by kbilsted** [comment]
@dolkensp  have you even read my posts? Try re-reading them before telling me that I'm defeated. Hint: I use the word *may* to broaden the discussion since no-one cares to read the theoretical papers. Specifically, I'm talking *semantics* so why are you talking *syntax* suddenly? Semantically the two constructs are different.

I'm withdrawing myself from this thread as people are too busy pointing fingers rather than discussing matter.

**Comment by AlexisFinn** [comment]
Wow, what the hell are you people using this equation for ? Trying to
calculate the trajectory to send a shuttle to Mars ? If you truly nead to
get absolute exact numbers out of a mathematical equation I suppose you're
just not looking in the right place.

The doc states: "PHP Depend can generate a large set of software metrics
from a given code base, *these values can be used to measure the quality of
a software project and they help to identify that parts of an application
where a refactoring should be applied*."

So the end goal is to accurately point out bits of code that appear to be
too complex and deserve to be looked into, not to give an exact npath
complexity result.

So far this as been working just fine for me (I don't care if the Npath is
500 or 5000, as long as it pops up as being maybe too much).

Once again if your goal is to obtain a perfectly exact computation of the
NPath complexity of a function then you're probably not looking in the
right place, because that is simply not what pdepend does.

2016-01-26 23:32 GMT+01:00 Kasper B. Graversen <notifications@github.com>:

**Comment by dolkensp** [comment]
@AlexisFinn

Well said - we should probably stop complaining when our calculator says that 5x5 = 20 either, because it's an ADDING device, and multiplication simply isn't what a calculator does.

If this bug truly isn't pdepend's purpose - then remove the metric. No point in having a metric that will spit out a number that's artificially inflated in some scenarios, but not in others. In that instance, the metric becomes completely useless, as you have nothing to compare against, so you might as well remove it

"Hey Joe, this method is scoring 5000 - should we check it?" "Nah - it's probably just a bunch of ternaries again" - meanwhile, it's actually a hot-path for the entire system because of some faulty branching logic.

It's a bit like going to see a doctor for a minor cut, and them amputating your foot as a result. "Their purpose isn't to save your foot, it's to stop the spread of infection - I don't care if the cut is a toe, or a leg, as long as the infection is gone" - yeah, great job.

**Comment by AlexisFinn** [comment]
"Hey Joe, this method is scoring 5000 - should we check it?" "Nah - it's
probably just a bunch of ternaries again"
Here lies the problem. Not hte metric, but the programmer who is just too
lazy to check things out by himself.
If your skill as a developper really depends that much on these metrics,
well I suppose you're out of luck, but it is truly a mistake to expect a
tool to do all the work for you, after all it's your code and your
responsibility do dertermine if the code is good or not.

In my previous post I gave an example of a simple validator function that
spits out and incredibly huge complexity result, but that code is fine.

If you want I can also find a bunch of real dirty code that will pass
through undetected.

You are the one to make the final decistion, not some tool, and if you
can't make that decision without the tool then find a better tool or stop
making decision, but laying the responsibily on the tool is just laughable.

2016-01-28 16:40 GMT+01:00 Peter Dolkens <notifications@github.com>:

**Comment by AlexisFinn** [comment]
"Hey Joe, this method is scoring 5000 - should we check it?" "Nah - it's probably just a bunch of ternaries again" - meanwhile, it's actually a hot-path for the entire system because of some faulty branching logic.
Yep, you should have looked at that. Pdepend warned you, you chose to ignore it because you figured the result must be wrong (and you're a bit lazy). Your fault, your responsibility. Not Pdepend's.

**Comment by Rarst** [comment]
> it just isn't a significant bug because it doesn't really have any impact on the purpose of the library

Since it completely subverts the _meaning_ of output from "this has high complexity" to "this uses syntax flavor which lib doesn't like" it has enormous impact on purpose of the library.

I agree that metrics like this are just indicators. When indicators react to something entirely different from what they are _supposed_ to indicate — they are broken.

**Comment by minow** [comment]
@AlexisFinn I disagree. When the tool is supposed to measure complexity, and provides you metrics that raises attention every time you run it, it's not fair to expect the developer to check for these spots every time the tool is run (or call developer lazy otherwise), it's just a huge inconvenience. Imagine this is reported in code that is perfectly fine and almost never changes in the project. Everytime you run pdepend you get this high score. Imagine there are 50 spots like this in a project. This just doesn't make sense. The tool should help you find spots with high complexity. This complexity should not be relative. 5000 is ok for code with many ?s, but otherwise it's bad and you should attend it. The tool then becomes useless IMHO. You could admit, that there are situations when this approach would lead developer to skip these reported high scores, if it's run for 100th time and it's always the same reason (bloated score because of ?s) and it happens in too many places. It's not laziness, it's just how mind works, I wouldn't blame programmer for being lazy, but the tool.

**Comment by dolkensp** [comment]
> Yep, you should have looked at that. Pdepend warned you, you chose to ignore it because you figured the result must be wrong (and you're a bit lazy). Your fault, your responsibility.

Do you check your house for a fire every time you burn toast and the fire alarm goes off too?

**Comment by AlexisFinn** [comment]
@minow Yes there are situations where this approach leads a dev to skip reported high scores, i've done it myself, and guess what I WAS WRONG! You can blame the tool all you like, it's just a tool, it's not an AI, it doesn't care! You're the one responsible, and if the tool doesn't do the job properly well then you're responsible for using the wrong tool (especially if you know it doesn't do the job).

**Comment by minow** [comment]
@AlexisFinn Doesn't it then make more sense then to fix this? People like to use these tools to help them in their daily tasks, they want to rely on the results, they want to get rid of need to do these calculations themselves. Are we now argueing that you should not trust any automation tool for any task and still do double checking manually? Would you argue the same way in case of e. g. automated testing? I mean is this a bug or not? Is the possible outcome 2 or 5? Is the tool suggesting you should replace all your ?s with if/elses? I think that is the point of this metrics, if the complexity is too high, try to lower it to have better comlexity. Isn't the code using either options equally right?

**Comment by AlexisFinn** [comment]
@minow Of course it makes more sense to fix it, but i'm not dong it, @manuelpichler as made it clear that he's not doing it, and I don't see anybody else forking the project and doing it, all I see is people complaining that the tool is warning them because the complexity is yyy whereas it should be warning them because the complexity is xxx wich is less than yyy (be of course it would still raise a warning).

And actually if you look at the source mathematical equation (http://dl.acm.org/citation.cfm?id=42379) used to compute NPath complexity, a ternary operator should give a result of 5. 
This may be a mistake, but pdepend correctly implements that equation and it's result.
It must be understood that NPath was devised to remplace McCabe’s metric as they were seen as flawed in many aspects, and the purpose of the equation is to point out complexity of code (albeit through computing the number of paths possible in a function).
So the end result is attained (wich is to point out difficult to understand/maintain code), and if there is some disagereement on the validity of the mathematical equation use to compute NPath complexity, well that doesn't depend on this library at all, this is merely an implementation of that equation.

**Comment by dolkensp** [comment]
@AlexisFinn - Can you copy/paste the paragraph in which it states this has not yet in this thread been revealed.

**Comment by yoosefi** [comment]
@AlexisFinn please provide a graph with 5 acyclical paths

**Comment by yoosefi** [comment]
three people so far have resorted to handwaving to some paper behind a paywall. enlighten me. that is all i've been asking for. draw me a stupid picture. please. i really do want to understand.

**Comment by yoosefi** [comment]
three people so far have resorted to handwaving to some paper behind a paywall.

the only argument anyone in favor of 5 paths has is "well, it's obvious. self-evident. who cares, anyway?"

ok. draw me a stupid picture, then. trace out these supposed paths. mspaint some dots with 5 lines connecting them that don't loop back, and label them. it would take you less time than writing all these essays.

**Comment by krisives** [comment]
Has anyone forked and fixed this issue yet if not I will do so just to make this die out of principle.

**Comment by lesliev** [comment]
@krisives that's what I suggested. But over on the reddit thread someone tried: 
https://www.reddit.com/r/programming/comments/42cf4n/on_researching_some_wacky_cyclomatic_complexity/czad4uq

According to them the tests, which are supposed make changes easy, are making changes impossible.

**Comment by AlexisFinn** [comment]
@dolkensp 
NPATH: A MEASURE OF EXECUTION PATH
COMPLEXITY AND ITS APPLICATIONS 
By BRIAN A. NEJMEH
Page 4 & 5:

> Syntax
> 
> (exprl) ? (expr2):(expr3)
> The semantics of the ? statement are as follows: The
> expression (expr I ) is evaluated first. If (expr 1 ) is
> nonzero (True), then the expression ( expr 2 ) is evaluated
> and returned as the result of the expression; otherwise,
> (expr3 ) is evaluated. Only one of (expr2) and
> ( expr 3 ) is evaluated. 
> 
> The acyclic execution path complexity for the ?
> operator is
> NP(?)=NP((exprl))+NP((expr2))
> +NP((expr3))+2.
> For our purposes, the ? operator can be treated similarly
> to the if -else statement. The 2 that is included
> in the NP(?) expression reflects the execution path
> complexity resulting from this statement (i.e., one path
> is traversed if (expr 1 ) is True, and another path traversed if (expr 1 ) is False
> 

So in (exprl) ? (expr2):(expr3) if Npath of expr1, 2 and 3 all  = 1:
1+1+1+2 = 5. 
The result is correct.
If the equation is wrong then the issue must be taken up with Brian E. Nejmeh, not with Pdepend developers.

**Comment by p2004a** [comment]
@AlexisFinn Seriously? You read only part of that paper?

Page 6:

    NP(expression)
        = number of && and I I operators in the expression.

I think it was end of discussion when @pygy quoted answer from Reddit thread. WTF. The NPATH complexity of expression *can be* 0!

**Comment by yoosefi** [comment]
I've emailed Nejmeh. Looking forward to his response.

**Comment by AlexisFinn** [comment]
@p2004a Well yes I hadn't looket at that part, and was supposing that any expression is evaluated to 1.
If NP((exp)) can equal 0 the it does make a bit more sense.

But then Npath of ($one && $two && $three) ? true : false; 
= 2 + 0 + 0 + 2 = 4. 

Ok, so that does make more sense in a way.

**Comment by dolkensp** [comment]
@yoosefi `$x = $a || $b || $c || $d;` should equal 3, assuming you have short-circuit evaluation of your binary operators.

You should be able to draw your own graph for it fairly easy with that assumption.

@AlexisFinn - Thanks for citing the paper - it seems between yourself and @p2004a, we have found where the confusion occurred.

I think it's fairly clear from the citations above, that the paper assumes short-circuiting behavior for binary operators, and that pdepend has made the incorrect assumption that NP(expr) >= 1.

**Comment by yoosefi** [comment]
Assuming short-circuiting is fallacious. Short circuiting is an optimization technique and is provided by languages/runtimes as syntactic sugar. As a pure operation, the components of `true || false || true` should be always be evaluated before the encompassing `OR`, giving 1 possible execution path every time. Thus there is no branching for a pure operation, and this paper is not about automata theory but instead catered to a specific category of languages.

Each branch increases the possible subsequent branches. You're telling me that && and || count as branches? So the "fix" for the NPath of 5 is to make it go even higher? Remind me what the point of this metric is again if we can't even use logical operators without accumulating astronomical "complexity" measures.

Let's distill the logical `OR` to a bitwise `OR`. The equivalent code is:
```
$a = 1;
$b = 0;
$c = 1;
$x = $a | $b | $c > 0;
```

Does this have NP 4? Are we now to include all math operations as complexity multipliers?

**Comment by dolkensp** [comment]
@yoosefi 

> As a pure operation, the components of true || false || true should always be evaluated before the encompassing OR

This is provably incorrect - http://php.net/manual/en/language.operators.logical.php

From my example, BINARY operators such as || and && are short-circuiting

From your example BITWISE operators, such as | and & are NOT short-circuiting.

They are absolutely NOT syntactic sugar - many languages provide it as a way to do things like

```if (collection.ContainsKey(someKey) && collection[someKey).IsGood == true)
{
    return collection[someKey].Name;
}```

**Comment by yoosefi** [comment]
@dolkensp you're saying that's the same as:

```
if (collection.ContainsKey(someKey)) {
    if(collection[someKey].IsGood == true) {
        return collection[someKey].Name;
    }
}
```

Which it's not.

**Comment by dolkensp** [comment]
@yoosefi yes it is - easy way to test:
```
function EchoAndReturnTrue() {
  echo "I return true";
  return true;
}

function EchoAndReturnFalse() {
  echo "I return false";
  return false;
}

function TestShortCircuiting() {
  if (EchoAndReturnFalse() && EchoAndReturnTrue()) {
    echo "I won't be run";
  }
}
```

You will only get a single line from this test script.

Nothing will be echoed

**Comment by yoosefi** [comment]
```
if (collection.ContainsKey(someKey)) {
    if(collection[someKey].IsGood == true) {
        return collection[someKey].Name;
    }
    return null;
}
```

Do that.

**Comment by senevoldsen90** [comment]
**Short**-circuiting logical operators do affect the number of paths possible. If the first expression is true in short circuiting logical and, the second expression is not evaluated. And since expressions may have side effect this impacts the number of acyclic paths, or NPATHs, through the code. An expression with 3 short circuiting AND has 4 total acyclic paths. Each statement reachable already confers 1 path, by the definitions in the paper the expressions adds an additional 3 giving a total of 4, which makes sense if you draw the graph:

http://graphviz.it/#/TgoBkYDJ/17

A more difficult issue, is that the paper seems to consider ternary operators as statements and not as expressions. This gives problems as you can obviously nest ternary operators, but also mix them with logical short circuiting operators. For instance, the algorithm does not consider this scenario:

```C
int x = (a > b ? c : (d ? e : f))
```
A naive reinterpretation would yield a NPATH of 4 although the number of acyclic paths are 3.

**Comment by krisives** [comment]
```
if (a > b) {
  if (d) {
    return e;
  } else {
    return f;
  }
} else {
  return c;
}
```

**Comment by tml** [comment]
Seems pretty clear at this point that this is a bug - one of you who actually uses Pdepend should fork it and patch it (maybe @yoosefi or @h0gar?) I don't actually use this tool, but glancing through the threads, it seems that it should be as simple as:

```
diff --git a/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php b/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php
index 809ce2f..48fe37b 100644
--- a/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php
+++ b/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php
@@ -211,10 +211,7 @@ class NPathComplexityAnalyzer extends AbstractCachingAnalyz
```er implements Analyze
         }

         foreach ($node->getChildren() as $child) {
-            if (($cn = $this->sumComplexity($child)) === '0') {
-                $cn = '1';
-            }
-            $npath = MathUtil::add($npath, $cn);
+            $npath = MathUtil::add($npath, $this->sumComplexity($child));
         }

         return MathUtil::mul($npath, $data);

**Comment by yoosefi** [comment]
i tried tracking down the behavior a while ago but it looks like it's spagetti'd out in multiple places.

**Comment by AlexisFinn** [comment]
Ok, so it appears to be a bug, but I still must point out that this in no way renders Npath unusable as the purpose is maintained.
In the paper Nejmeh points out several times that NPATH is meant as a guidance to point out potential complexity/maintenance issues, but only as a guidance, it doen't point out godd or bad code, it just points out code that may warrant a closer look.

> NPATH can be used in determining
the level of review/inspection of a function. Functions
with high (i.e., in the top 25 percent) NPATH
values are candidates for thorough review and/or inspection.
Thus, NPATH is aiding in the process of deciding
which functions should be thoroughly reviewed
and/or inspected.
[...]
> Any decision to allocate inspections, testing, or design
effort based on NPATH must also take into account
the criticality of the function: Whereas even moderately
high NPATH values in a heavily used function
would identify the function for thorough inspection
and testing as well as possibly redesign, a noncritical
function of similar NPATH complexity might not
warrant the same level of attention; and
factors other than complexity impact on software
quality: Requirements volatility, software development
environment, developer experience, reuse, and
the use of code generators all impact on software
quality.
Thus, the use of NPATH cannot provide absolute principles
for software development, but is a useful adjunct
to traditional and intuitive measures of software complexity.
[...]
Although strategies to reduce the NPATH complexity
of functions are important, care must be taken not to
distort the logical clarity of the software by applying a
strategy to reduce the complexity of functions. That is,
there is a point of diminishing return beyond which a
further attempt at reduction of complexity distorts the
logical clarity of the system structure.

So whereas the number of acyclic execution paths is being slightly accurate is kind of bothering, the point is that: 
1) it still accurately points out code that needs reviewing.
2) it still scales proportionally with the ressources allocated to functional testing.
3) it still serves as a usefull indicator as far as design patterns go, 
and as a side note I think this bug makes it even better in my opinion as it is basically telling you to not use ternary operators.

PS: @senevoldsen90 Unknown to most there is actually a tenth circle in hell made specifically for people who use nested ternary operators.

**Comment by tml** [comment]
Generally, I agree with @AlexisFinn that this is more of an "annoyance" bug than a fundamental flaw; however, it's possible that some organizations have adopted a ridiculous rule such as "thou shalt not commit code with an NPATH greater than X" - I've heard dumber things - and this bug might be causing someone to not get perfectly acceptable code past such an arbitrary gate. On the other hand:

> and as a side note I think this bug makes it even better in my opinion as it is basically telling you to not use ternary operators.

I think that this is the problem some people have with this bug. There's no reason not to use the ternary operator if it is the clearest expression of the intent of the code.

**Comment by AlexisFinn** [comment]
@tml Ternary operators are never the clearest expression,
Ex:
`$result = ($test > 0) ? 'more': 'less';`
Pretty simple and straightforward, no problem there.
But when compared to this
```
if ($test > 0) {
    $result = 'more';
} else {
    $result = 'less';
}
```
Well the if/else syntax while being much less compact is practically written in plain english, and there's a good chance that even someone with no experience in programming would understand it.
Now show the ternary notation to that same person and he will undoubtedly be dumbfounded.

Also i've just way too often seen things like 
`
$result = ($bigger > $whatnot && $smaller == $notBigger)?($smaller !== 0 && isset($bigger))?$bigger/$smaller:$bigger*round($whatnot, 2):($whatnot < 0)?0:$bigger;
`
And this is actually an easy one, believe it or not but i've seen triple nested ternary operators with even longer and more complex conditions.

So whereas I don't really have any complain about a properly used ternary condition, the frequency at which i've come across totally improper use of them and the incredible headache it gives me each time makes me strongly believe that not only should they definitely warrant a higher complexity values, making it more likely that someone will review there usage, but they should (unfortunately) be discouraged altogether.

**Comment by senevoldsen90** [comment]
I can follow the reasoning of any who propose that ternary expressions have some inherent greater complexity than if-than-else. And if you believe that, then it certainly makes sense to have some complexity metric for your team that reports them as more complex. Maybe that complexity metric is something completely else, or some variant of NPATH; but that complexity is not NPATH. When someone asks for the NPATH they have something specific in mind: the number of acyclic paths through the code. We may _believe_ that ternary expressions, or gotos, are more complex than other forms of expression, that belief does not change their NPATH values.

**Comment by AlexisFinn** [comment]
@senevoldsen90 Agreed, no matter how I feel about ternary operators this is still a bug in the implementation of the equation, I was just pointing out that the NPATH computation was never meant to be used in and of itself, but as tool (amongst others) for the purpose of 
1) pinpointing potentially excessive complexity in the code warranting a code review, 
2) evaluating allocation of testing ressources,
3) helping to come up with better design patterns for classes and functions. 
And in none of these aspects do I feel that the tool fails.
As a matter of fact Nejmeh himself stresses the point that it is meant to be used as a guide for human intervention, and even points out that it can easily be fooled by duplicating a test like such:
```
function {
  if (a == b) {
    do something;
  }
  if (a == b) {
    do something else;
  }
}
```
NPATH will not recognise these two paths as being the same and spit out a wrong result (4 instead of 3).

But as far as the implementation of the NPATH equation goes, yes, I agree it's a bug.

**Comment by senevoldsen90** [comment]
@AlexisFinn I think we are very much in agreement about how such a tool should be used. Besides the bug itself, one worry remains. Would it generate too many false positives because of the increased exponential growth? Consider some code that does some initialising at the top:
```
$option1 = f($a) ? $a : $b;
$option2 = g($b) ? $c : $d;
$option3 = g($option1) : true : false;
// Remainder has, say, NPATH = 10.
```

With NPATH of 5 for the ternaries it would give: 5 x 5 x 5 x 10 = 1250
Rewritten as if-than-else: 3 x 3 x 3 x 10 = 270
Rewritten as assignment + if-than: 2 x 2 x 2 x 10 = 80

That's more than a 4 times difference. With another ternary, it would be close to 8 times different. But, I'll be honest, I don't if that would generate significant "false positives". But, then again, it's not like there is a choice until it is fixed.

**Comment by AlexisFinn** [comment]
@senevoldsen90 that is in my opinion a false problem, either you can of course rewrite this (wich in my opinion would be useless), or you could simply review the code and decide that NPATH can be ignored in this case besause the code is fine.
In my previous post https://github.com/pdepend/pdepend/issues/158#issuecomment-163967227 I gave a real-life example of a formValidation function wich spits out an incredibly high NPATH complexity value, but truly doesn't warrant any changes.
And that's without using ternary operators.
People seem to think that high NPATH = bad code, and low NPATH = good code, and that's simply not true.
Another quick example:
```
if (true == false) {
    ...
}
if (0 > 1) {
    ...
}
...
```
Well Npath in this case is 4 (wich of course is stupid), but thats a pretty low NPATH so is that code good ?

**Comment by senevoldsen90** [comment]
@AlexisFinn In retrospect my use of "false positive" was a poor choice of words. I did not mean to bring us back to the binary evaluation of bad code or good code.

What I meant was, that it might misdirect disproportionally more of limited attention towards code that do not really need it. I have only limited experience with the methodology of using these tools, so I'll defer to your experience, saying it's not an issue.

**Comment by AlexisFinn** [comment]
@senevoldsen90 Lol, thank you but defering to me might not be the best Idea, but the worst idea is deferring to a metric like NPATH. It's purpose is to point you in the right direction, as for misdirecting attention, I suppose thats somewhat true, but thats why other indicators such as Cyclomatic complexity, copy-paste detector and tools such as code-sniffer and mess-detector should be properly configured and used alongside NPATH so as to circumvent any of one tool's limitations.

But no matter what, nothing will replace a proper review from a human being, I've seen horrendous code that would never trigger an NPATH warning, and perfectly fine code that spits out huge NPATH complexity.

Then you must wonder what is the purpose of this metric, what are you trying to achieve by using it?
If your goal is to reduce NPATH at all costs, well that's easy, just wipe all your code and voila !, but most likely you goal is to produce cleaner, easily understandable and easily maintainable code.
And sometimes rewriting you functions to lower the NPATH complexity will actually increase the complexity of the code, making it harder to understand and maintain, and basically defeating the whole purpose.

In the end it's up to the user to decide wether the NPATH warning is pertinent or not, and the only way to do that is to look at the code with a full understanding of it's prupose while keeping in mind the goal you are trying to achieve.

Basically defer to yourself.

**Comment by yoosefi** [comment]
lmao end my life already

**Comment by yoosefi** [comment]
why is everyone taking nejmeh as gospel? he has some fake degree from some unaccredited uni.

**Comment by offsides** [comment]
A but late to the party, but I've only just started using phpmd (which uses pdepend on the back end) and was looking into the whole NPath question when I found this.  I use a LOT of ternary operators, and have functions with NPath values in the billions because of it, so looking into the root cause piqued my interest.  I believe I've figured out the root cause of the problem, which essentially is poor logic in the original definition.  I haven't seen the original paper so I can't confirm that the NPath value for the ternary operator is listed there the same way, but from what I've seen the formula used is:

NP(?) = NP(expr1) + NP(expr2) + NP(expr3) + 2

Now, the implementation forces NP(exprN) to have minimum value of 1, leading to a minimum total value of 5.  But looking at the formula, I believe that the whole '+2' thing was based on the concept that NP(exprN) could return a minimum value of 0, not 1, as 3 simple values would return a total NP value of 2 rather than 5.  But that still doesn't resolve the problem completely, as it wouldn't account properly for nested ternary operators like this:

NP (x?a:b) = 2 (where NP of x, a and b = 0)
NP (x?a:(y?b:c)) = 4 (again, where NP of x, y, a, b and c = 0)

This is, again, because the '+2' was designed with the assumption that each 'branch' (return value) should have a minimum value of 1, and if the NP values were 0  for no branches it would add up correctly.  But it doesn't, as NP values have a minimum of 1 not 0.  And there's actually a reason for that: since the total NPath value is the product of each individual NP value, and not the sum, an NP value of 0 anywhere in the formula would result in a total NPath value of 0, which is obviously incorrect for even the simplest of functions.  So the formula needs to be adjusted to take into account minimum values of 1 rather than 0.

To resolve this issue, I would suggest (and I have no idea where or how to make it 'official') that the NP(?) formula be changed to this:

NP(expr1 ? expr2 : expr3) = NP(expr1) + NP(expr2) + NP(expr3) - 1

Now, using the nested examples above, but with an NP value of 1 for each term, we get the following:

NP (x?a:b) = 2 (where NP of x, a and b = 1)
NP (x?a:(y?b:c)) = 3 (again, where NP of x, y, a, b and c = 1)
and extended:
NP (w?a:(x?b:(y?c:(z?d:e)))) = 5 (where NP of all terms = 1), which is just what you'd expect since there's 5 possible return values.

If anyone knows how to submit this to whoever for consideration as a change to the spec, I'd be happy to do so.  I'd love to see the original paper to see what it says, but I'm not sure I want to spend money for it...

**Comment by offsides** [comment]
Having just acquired a copy of the original paper I can say for certain that the implementation of the NPath calculation for the ternary operator is incorrect.  Additionally, there is another bug in the implementation that essentially causes it to ignore the NPath value of the conditional expression within the ternary operator.  I will file a new issue covering both pieces of it including various documentary evidence once I compile it.

**Comment by yoosefi** [comment]
@offsides nice research. please comment the new issue number here. thanks

**Comment by offsides** [comment]
New issue is #261

**Comment by yoosefi** [comment]
Why was it closed without any comments? Did you do that?

**Comment by tml** [comment]
This issue has been closed for almost 2 years; manuelpichler closed it. Issue #261 remains open.

**Comment by yoosefi** [comment]
@tml sorry i misread the log

**Comment by felixkk** [comment]
Been a real pleasure to read this from the start xD
